src/Functions/ztest.cpp

```diff
@@ -0,0 +1,199 @@
+        }
+
+        String getName() const override
+        {
+            return name;
+        }
+
+        size_t getNumberOfArguments() const override { return 6; }
+        ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {5, 6}; }
+
+        bool useDefaultImplementationForNulls() const override { return false; }
+        bool useDefaultImplementationForConstants() const override { return true; }
+        bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+        static DataTypePtr getReturnType()
+        {
+            DataTypes types
+            {
+                std::make_shared<DataTypeNumber<Float64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+                std::make_shared<DataTypeNumber<Float64>>(),
+            };
+
+            Strings names
+            {
+                "z_statistic",
+                "p_value",
+                "confidence_interval_low",
+                "confidence_interval_high"
+            };
+
+            return std::make_shared<DataTypeTuple>(
+                std::move(types),
+                std::move(names)
+            );
+        }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+        {
+
+            for (size_t i = 0; i < 4; ++i)
+            {
+                if (!isNativeInteger(arguments[i]))
+                {
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The {}th Argument of function {} must be a integer or unsigned integer.", i + 1, getName());
+                }
+            }
+
+            if (!isFloat(arguments[4]))
+            {
+                throw Exception{"The fifth argument of function " + getName() + " should be a float, illegal type: " + arguments[4]->getName(), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
+            }
+
+            if (!isString(arguments[5]))
+            {
+                throw Exception{"The sixth argument of function " + getName() + " should be a string, illegal type: " + arguments[5]->getName(), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
+            }
+
+            return getReturnType();
+        }
+
+
+        static void nan(MutableColumnPtr & to)
+        {
+            const Float64 nan = std::numeric_limits<Float64>::quiet_NaN();
+            Tuple tuple({nan, nan, nan, nan});
+            to->insert(tuple);
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            auto non_const_arguments = arguments;
+            for (auto & argument : non_const_arguments)
+                argument.column = argument.column->convertToFullColumnIfConst();
+
+            MutableColumnPtr to{getReturnType()->createColumn()};
+            to->reserve(input_rows_count);
```

Comment:
Result of this function is tuple column with 4 float columns, consider to allocate 4 float columns, insert data into them, then make tuple from them.
That way we will avoid a lot of unnecessary indirections, creating unnecessary tuples.